### PR TITLE
Console driver cleanup

### PR DESCRIPTION
**Pull request by abrestic-rivos**
Introduce a `Console` object for implementing `print!` and `println!` that can have the writer set at runtime via a `ConsoleWriter` trait object (patches 1 & 2). This allows us to switch from SBI to UART console, and to use the console outside of the binary crates. Finally, clean up the UART driver by moving it to the `drivers` crate (patch 3) and probe for it from the device tree so that we don't have to use hard-coded addresses (patches 4 & 5).

### CODE REVIEW

File: s-mode-utils/src/print.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2021 by Rivos Inc.
```
Comment (dgreid) [inline s-mode-utils/src/print.rs:1]: 2022

File: s-mode-utils/src/sbi_console.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
+// Copyright (c) 2021 by Rivos Inc.
```
Comment (dgreid) [inline s-mode-utils/src/sbi_console.rs:1]: 2022

File: drivers/src/uart.rs [hunk 3]
```diff
@@ -15,17 +36,37 @@ pub struct UartDriver {
...
+        {
+            return Err(Error::InvalidRegisterLocation);
+        }
+        // Safety: We trust that the device tree accurately described the location of the UART.
+        unsafe {
+            mem_map
+                .add_mmio_region(DeviceMemType::Uart, RawAddr::supervisor(base_address), len)
+                .map_err(Error::AddingMmioRegion)
+        }?;
+        // Unwrap ok, we've already rified that base_address is non-NULL.
```
Comment (dgreid) [inline drivers/src/uart.rs:63]: verified?

**Review by dgreid — APPROVED** [review]
way better, thanks!

File: s-mode-utils/src/print.rs (hunk 1 above)
Comment (abrestic-rivos) [reply s-mode-utils/src/print.rs]: Copy-pasta. Done.

File: s-mode-utils/src/sbi_console.rs (hunk 2 above)
Comment (abrestic-rivos) [reply s-mode-utils/src/sbi_console.rs]: Done.

File: drivers/src/uart.rs (hunk 3 above)
Comment (abrestic-rivos) [reply drivers/src/uart.rs]: Done.

**Pull request merged**